### PR TITLE
Upgrade to purity inference

### PR DESCRIPTION
**Pull request by lukewilliamboswell**

### CODE REVIEW

File: examples/echo.roc [hunk 1]
```diff
@@ -1,41 +1,35 @@
...
+app [main!] { pf: platform "../platform/main.roc" }
 
 import pf.Stdin
 import pf.Stdout
 
-main =
-    Stdout.line! "Shout into this cave and hear the echo!"
+main! = \{} -> tick! {}
 
-    Task.loop {} tick
```
[the remark this answers is not in the document]
Comment (JanCVanB) [reply examples/echo.roc:9]: Ah, now I see the recursion on line 12. I infer that plain recursion is the intended replacement.

File: examples/echo.roc [hunk 2]
```diff
@@ -1,41 +1,35 @@
-app [main] { pf: platform "../platform/main.roc" }
+app [main!] { pf: platform "../platform/main.roc" }
 
 import pf.Stdin
 import pf.Stdout
 
-main =
-    Stdout.line! "Shout into this cave and hear the echo!"
```
[the remark this answers is not in the document]
Comment (lukewilliamboswell) [reply examples/echo.roc:7]: Restored in 7fe35963a

**Comment by smores56** [comment]
It seems like our `flake.nix` needs to have `valgrind` added to the `sharedInputs` array. I tried pushing a commit doing this, but didn't have write perms. If I'm not overstepping, could I request write access for this repo so that I can make fixes like this? Feel free to say no.

**Comment by Anton-4** [comment]
I did not change the permissions but you do appear to have write permission now @smores56.

**Comment by Anton-4** [comment]
No need to worry about valgrind on nix, the next step [is this](https://roc.zulipchat.com/#narrow/channel/316715-contributing/topic/basic-cli.20purity.20inference/near/486160194).

File: Cargo.toml [hunk 3]
```diff
@@ -7,7 +7,7 @@ lto = true
 strip = "debuginfo"
 # You can comment this out if you hit a segmentation fault similar to the one in see issue github.com/roc-lang/roc/issues/6121
 # Setting this to 1 should improve execution speed by making things easier to optimize for LLVM.
-codegen-units = 1
+# codegen-units = 1
```
Comment (lukewilliamboswell) [inline Cargo.toml:10]: I'm not sure what's going on here. I don't remember changing this

File: ci/all_tests.sh [hunk 4]
```diff
@@ -67,6 +67,12 @@ for roc_file in $EXAMPLES_DIR*.roc; do
 
     roc_file_only="$(basename "$roc_file")"
     no_ext_name=${roc_file_only%.*}
+
+    # not used, leaving here for future reference if we want to run valgrind on an example
+    # if [ "$no_ext_name" == "args" ] && command -v valgrind &> /dev/null; then
+    #     valgrind $EXAMPLES_DIR/args argument
+    # fi
```
Comment (lukewilliamboswell) [inline ci/all_tests.sh:74]: Leaving this here for future... we went to all the effort to use valgrind

**Review by bhansconnect — APPROVED** [review]
Long time coming!!!

File: platform/Arg.roc [hunk 5]
```diff
@@ -1,106 +1,14 @@
...
-            Task.err (Exit 1 "")
+list! : {} => List Str
+list! = \{} ->
+    PlatformTasks.args! {}
+
+# TODO
+# I deleted the Arg parser as a workaround for issues it was causing when upgrading to purity inference.
+# The plan is for people so use Weaver or simmilar in future. We can discuss putting it back in if we still
+# want the "batteries included" experience, but either way that will need to be upgraded/tested with
+# purity-inference.
```
Comment (lukewilliamboswell) [inline platform/Arg.roc:14]: TODO note... I can remove this in this or follow up, I left this here for @Anton-4

File: Cargo.toml (hunk 3 above)
Comment (bhansconnect) [reply Cargo.toml:10]: Oh, probably should update that back to `codegen-units = 1` in a follow up

**Pull request merged**